Foreign definitions for modules Common, CodeUnits and Unsafe

postsolar opened the pull request:
These come together because they are needed to be able to reproduce a minimal tests' subset run. Imports imply/depend on merging of the IrRegex PR, as well as SRFI N152 PR for `purescm/purescm`.

postsolar committed:
Update workflows

postsolar committed:
Add a note about lone surrogates in Test.Data.String.CodePoints

postsolar committed:
Add spago hash

postsolar committed:
Comment out tests outside the scope of this PR

postsolar committed:
Add foreign definitions for `CodeUnits`, `Common` and `Unsafe`

These come together because they were all needed to run the minimal test subset.

anttih commented on src/Data/String/CodeUnits.ss:
```diff
@@ -0,0 +1,128 @@
+        (lambda (x)
+          (lambda (startAt)
+            (lambda (s)
+              (let ((clamped (clamp-to-length s (+ startAt (string-length x)))))
```
I don't understand this clamping or why we need to do this. Shouldn't we just pass the `startAt` as-is to the `string-contains-right`?

anttih replied:
Is it because `startAt` can be negative?

postsolar replied:
Clamping in general is needed because the library throws when index is outside the bounds. The weird `(+ startAt (string-length x))` thing is there because... it's easier to explain with an example from tests:
```
  assertEqual
    { actual: SCU.lastIndexOf' (Pattern "bc") 1 "abcd"
    , expected: Just 1
    }
```
I think it's more intuitive to expect it to return Nothing because if you start at index 1 and move left you won't ever get "bc", so you see the assumptions the JS/PS API makes about indexing here. That's why I add the length of the target pattern `x` to the starting index.

anttih replied:
Ok so `x` is not the whole string, it's the pattern. I missed that.

anttih commented on src/Data/String/CodeUnits.ss:
```diff
@@ -0,0 +1,128 @@
+            (cons "before" (srfi:152:string-take s clamped))
+            (cons "after" (srfi:152:string-drop s clamped)))))))
+
+  (define clamp-to-length
```
Should this be called `clamp-index` or something like that since what this is really doing is making sure the index is valid?

anttih commented on src/Data/String/CodeUnits.ss:
```diff
@@ -0,0 +1,128 @@
+  (define splitAt
+    (lambda (i)
+      (lambda (s)
+        (let ((clamped (clamp-to-length s i)))
```
Maybe name this binding so that it's clear it's an index. `clamped` is a bit vague IMO.

anttih commented on src/Data/String/Unsafe.ss:
```diff
@@ -0,0 +1,28 @@
+
+    (define char
+      (lambda (s)
+        (if (= (string-length s) 1)
+          (string-ref s 0)
+          (raise-continuable
+            (make-message-condition
+              "Data.String.Unsafe.char: Expected string of length 1.")))))
```
Can't we just call `string-ref` here and not do the bounds check?

postsolar replied:
We would then, at least, need to catch it anyways to format the error. But that aside, the main reason is that the PS function is explicitly like this: a string of length 1, no more no less, else I throw. This here is direct mirroring of the JS implementation.

anttih replied:
I think the only reason the JS version of both of these functions throw an exception "manually"  is beause `String.prototype.charAt` itself won't throw an exception:

```
> "asdfa".charAt(100)
''
> "".charAt(0)
''
```

The point of the `Unsafe` module is to be unsafe, so I think we should do the bare minimum to return the correct value and not do any special handling. IMO we should remove all catching from this module _and_ also accept that `char` will return a value even if the string has a length of more than 1. `string-ref` will itself throw an exception if it can't return a `Char` at the given index. I would not care about the error message string.

anttih reviewed (commented):
Small stylistic nitpick: could you use square brackets in the `let` form for the binding pairs?

anttih reviewed (commented):
With this we also arrive at the spot where we need to figure out how to deal with the fact that JS strings are UTF-16 encoded code units and Chez uses unicode code points:

```
> "😁a😁😁".charAt(1)
'\ude01'
```

```
> (string-ref "😁a😁😁" 1)
#\a
```

For example the `CodeUnit` module now deals with code points rather than code units. We could actually make it work with code units by turning the string to a bytevector with `string->utf16` and indexing into that.

postsolar committed:
Brackets + clearer namings

postsolar committed:
Remove all error handling from `Data.String.Unsafe` FFI

postsolar commented:
> With this we also arrive at the spot where we need to figure out how to deal with the fact that JS strings are UTF-16 encoded code units and Chez uses unicode code points:

Something like this?

```scheme
  (define _charAt
    (lambda (just)
      (lambda (nothing)
        (lambda (i)
          (lambda (s)
            (if (< i (string-length s))
              (let
                ([cus (bvs:string->utf16 s)]
                 [v (bvs:make-bytevector 2)]
                 [ix (* i 2)]
                 [tx (make-transcoder (utf-16-codec))])
                (bvs:bytevector-u16-set!
                  v
                  0
                  (bvs:bytevector-u16-ref cus ix (bvs:native-endianness))
                  (bvs:native-endianness))
                (just (string-ref (bytevector->string v tx) 0)))
              nothing))))))

```

postsolar committed:
Bring back tests for Data.String.CodeUnits

Doesn't make sense for them to be commented out if the module they're testing is under review

postsolar committed:
Data.String.CodeUnits: Implement `_charAt`, `_toChar`, `length` and `countPrefix` in terms of CUs

This temporarily makes tests fail because the rest of functions don't expect `length` to be returning the values it's returning now

postsolar committed:
Fix bounds checking in `Data.String.CodeUnits._charAt`

It was doing it wrt the unicode string's length rather than utf-16 bytevector length

postsolar committed:
Off-by-one error in countPrefix

postsolar committed:
Implement the rest of `Data.String.CodeUnits` in terms of CUs

postsolar committed:
Fix indexOf, lastIndexOf, indexOfStartingAt and lastIndexOfStartingAt

In previous commit I forgot to make them operate on CUs

postsolar committed:
Revert "Fix indexOf, lastIndexOf, indexOfStartingAt and lastIndexOfStartingAt"

Actually they were just fine.

This reverts commit 7e2715d038e713f8c604275f3a46012c364d0726.

postsolar commented:
Ok I think it's ready.

postsolar commented:
A note to myself: for SRFI 152, potentially there's a huge performance gain in using `substring` primitive instead of `take` and especially `drop`.
